python/ray/data/impl/stats.py

```diff
@@ -0,0 +1,239 @@
+                        fmt(np.mean([e.cpu_time_s for e in exec_stats])),
+                        fmt(sum([e.cpu_time_s for e in exec_stats]))))
+
+        output_num_rows = [
+            m.num_rows for m in blocks if m.num_rows is not None
+        ]
+        if output_num_rows:
+            out += ("* Output num rows: {} min, {} max, {} mean, {} total\n".
+                    format(
+                        min(output_num_rows), max(output_num_rows),
+                        int(np.mean(output_num_rows)), sum(output_num_rows)))
+
+        output_size_bytes = [
+            m.size_bytes for m in blocks if m.size_bytes is not None
+        ]
+        if output_size_bytes:
+            out += ("* Output size bytes: {} min, {} max, {} mean, {} total\n".
+                    format(
+                        min(output_size_bytes), max(output_size_bytes),
+                        int(np.mean(output_size_bytes)),
+                        sum(output_size_bytes)))
+
+        if exec_stats:
+            node_counts = collections.defaultdict(int)
+            for s in exec_stats:
+                node_counts[s.node_id] += 1
+            out += (
+                "* Tasks per node: {} min, {} max, {} mean; {} nodes used\n".
+                format(
+                    min(node_counts.values()), max(node_counts.values()),
+                    int(np.mean(list(node_counts.values()))),
+                    len(node_counts)))
+
+        return out
+
+
+class DatasetPipelineStats:
+    """Holds the execution times for a pipeline of Datasets."""
+
+    def __init__(self, *, max_history: int = 3):
+        """Create a dataset pipeline stats object.
+
+        Args:
+            max_history: The max number of dataset window stats to track.
+        """
+        self.max_history: int = max_history
+        self.history_buffer: List[Tuple[int, DatasetStats]] = []
+        self.count = 0
+        self.wait_time_s = []
+
+        # Iteration stats, filled out if the user iterates over the pipeline.
+        self.iter_wait_s: float = 0
+        self.iter_user_s: float = 0
+        self.iter_total_s: float = 0
+
+    def add(self, stats: DatasetStats) -> None:
+        """Called to add stats for a newly computed window."""
+        self.history_buffer.append((self.count, stats))
+        if len(self.history_buffer) > self.max_history:
+            self.history_buffer.pop(0)
+        self.count += 1
+
+    def summary_string(self) -> str:
+        """Return a human-readable summary of this pipeline's stats."""
+        already_printed = set()
+        out = ""
+        for i, stats in self.history_buffer:
+            out += "== Pipeline Window {} ==\n".format(i)
+            out += stats.summary_string(already_printed)
+            out += "\n"
+        out += "##### Overall Pipeline Time Breakdown #####\n"
+        # Drop the first sample since there's no pipelining there.
+        wait_time_s = self.wait_time_s[1:]
```

Comment:
This might surprise users, who may be trying to figure out what their pipeline is slow when it ends up being unavoidable latency of waiting for the first dataset in the pipeline. Could we add a flag to `.summary_string()` for including/excluding the first dataset's wait time?

Reply:
Added.

Reply:
Looks like the `exclude_first_window` argument isn't used, probably forgot an `if` statement here.

Reply:
Oh :facepalm: , fixed.